druid/src/main/java/org/apache/calcite/adapter/druid/DruidRules.java

```diff
@@ -604,30 +798,36 @@ private static boolean validSortLimit(Sort sort, DruidQuery query) {
         // offset not supported by Druid
         return false;
       }
-      if (query.getTopNode() instanceof Aggregate) {
-        final Aggregate topAgg = (Aggregate) query.getTopNode();
-        final ImmutableBitSet.Builder positionsReferenced = ImmutableBitSet.builder();
-        for (RelFieldCollation col : sort.collation.getFieldCollations()) {
-          int idx = col.getFieldIndex();
-          if (idx >= topAgg.getGroupCount()) {
-            continue;
-          }
-          // has the indexes of the columns used for sorts
-          positionsReferenced.set(topAgg.getGroupSet().nth(idx));
-        }
-        // Case it is a timeseries query
-        if (checkIsFlooringTimestampRefOnQuery(topAgg.getGroupSet(), topAgg.getInput(), query)
-            && topAgg.getGroupCount() == 1) {
-          // do not push if it has a limit or more than one sort key or we have sort by
-          // metric/dimension
-          return !RelOptUtil.isLimit(sort) && sort.collation.getFieldCollations().size() == 1
-              && checkTimestampRefOnQuery(positionsReferenced.build(), topAgg.getInput(), query);
+      // Use a different logic to push down Sort RelNode because the top node could be a Project now
+      RelNode topNode = query.getTopNode();
+      Aggregate topAgg;
+      if (topNode instanceof Project && ((Project) topNode).getInput() instanceof Aggregate) {
+        topAgg = (Aggregate) ((Project) topNode).getInput();
+      } else if (topNode instanceof Aggregate) {
+        topAgg = (Aggregate) topNode;
+      } else {
+        // If it is going to be a Druid select operator, we push the limit if
+        // it does not contain a sort specification (required by Druid)
+        return RelOptUtil.isPureLimit(sort);
+      }
+      final ImmutableBitSet.Builder positionsReferenced = ImmutableBitSet.builder();
+      for (RelFieldCollation col : sort.collation.getFieldCollations()) {
+        int idx = col.getFieldIndex();
+        if (idx >= topAgg.getGroupCount()) {
```

Comment:
I am confused here. If ```topNode``` is a Project (L804), then the Sort operator will reference the Project, not the Aggregate operator. Thus, I think now we need more complex checking here, as Project operator might be doing reordering of columns, computing expressions, etc.